### ISSUE TITLE
Implement tests for Streaming Receiver in PeekLock mode

### ISSUE THREAD

**Issue by AlexGhiondea** [Azure/azure-service-bus-node#101]
- [x] With auto-complete enabled + no message settling by the user should remove the message from the Queue/Subscription
- [x] With auto-complete disabled, manual completion by the user should remove the message from the Queue/Subscription
- [x] With auto-complete disabled, dead lettering the message results in not getting the same message again from queue/subscription. The message is then gotten only from the dead letter queue
- [x] With auto-complete disabled, abandoning a message results in getting the same message from queue/subscription with incremented delivery count. After 10 times, you get it no more. You can then only get it from the dead letter queue
- [x] With auto-complete enabled + manual 
   - [x] completion
   - [x] abandon
   - [x] defer 
   - [x] deadletter 
in Queue/Subscription by the user should not result in SDK trying to complete it again and no errors should be thrown
- [x] Second Streaming Receiver call should fail if the first one is not stopped
- [ ] Settling a message twice should throw error. (Blocked by #74)
- [x] With auto-complete disabled, deferring a message results in not getting the same message again from queue/subscription. The message is then gotten using `receiveDefferedMessages`
- [x] With auto-complete disabled, no settling of the message results in getting the same message again from queue/subscription
- [ ] Test `maxConcurrentCalls`. When set to 1 (default value), next message is received only after the first one is settled. Test with non default value as well. (Tracked in #179)

**Comment by HarshaNalluru** [comment]
**All the tests here are done except** 
- Settling a message twice should throw error. (Blocked by #74)
------------------------------------
**Issues I've faced while working on tests:** 
- Receiving messages from a deadletter gives a timeout ( happens infrequently )
   - If working with a ServiceBus namespace gives the timeout errors, using a different ServiceBus (different namespace) helped getting rid of those timeouts.
- Clearing the messages using `receive()` did not work(second `receive()` didn't get invoked). So, cleared the messages using `receiveBatch()` instead in some of the tests.
- Some of the tests only work with Partitioned Queues and Partitioned Subscriptions.
---------------------------
**I've checked the list of tests for `ReceiveBatch`(#100). The following are the ones are not covered in this issue. Do we need tests similar to them in this issue ?** 
- Multiple sequential ReceiveBatch using Queue/Subscription, ensure result from one is not affected by the other
- Cases for which we are not yet sure what the result should be, but we should definitely cover: 
    - `Complete/Abandon/Deadletter/Defer` a message received from `recevieDefferedMessage`, what is the expectation?
    - `Complete/Abandon/Deadletter/Defer` a message received from the dead letter, what is the expectation?
---------------------------
@AlexGhiondea @ramya-rao-a @ShivangiReja @ramya0820

**Comment by ramya-rao-a** [comment]
- Second Streaming Receiver call should fail if the first one is not stopped
The test does check the error message in the catch() block. But this only guarantees that the error message is the expected one assuming that the error was indeed thrown. But there is nothing in the test that tests that the error indeed was thrown.

- Manual abandon() and defer() 
We should test that the delivery count gets updated in these cases

**Comment by HarshaNalluru** [comment]
> Comments on existing test cases:

**Second Streaming Receiver call should fail if the first one is not stopped**
- Great catch, I'll keep a flag and add a check after the catch statement.

**Manual abandon() and defer()
We should test that the delivery count gets updated in these cases**
- Will update after both the PRs(#171 and #170) are merged.
-----------------------------------------------

> We know that receiving deferred messages from unpartitioned queues/subscriptions are not working. Did you encounter anything else?

Even the `AutoComplete removes the message from Subscription` keeps failing at the unpartitioned queues and I've seen some other tests keep failing. I'll re-check and make the list of the tests which don't work and will see if we can fix by adding more delays.

-----------------------------------------------
> Complete/Abandon/Deadletter/Defer a message received from the dead letter, what is the expectation?
> We could do this. We did this already in batching receiver.

Okay, will do these along with the above comments on "Second Streaming Receiver call " and "Manual abandon() and defer()" after both the PRs(#171 and #170) are merged.

**Comment by ramya-rao-a** [comment]
@HarshaNalluru Dont wait for #170 to be merged yet. Let's park that for now. I have some doubts in the expected behavior there.

And yes, please do make a list of tests that do not work for unparitioned queues/subscriptions

**Comment by HarshaNalluru** [comment]
Oh okay. I see #171 is merged already. So, will work on these on a different branch.

**Comment by ramya-rao-a** [comment]
Closing this issue as the pending tasks are being tracked in #74, #179 and #180
